core/src/mock_alpenglow_consensus.rs

```diff
@@ -0,0 +1,651 @@
+        trace!(
+            "{}: preparing to receive for slot {slot}",
+            self.cluster_info.id()
+        );
+        let staked_nodes = self.epoch_specs.current_epoch_staked_nodes();
+
+        let mut state = get_state_for_slot(self.state.as_slice(), slot)
+            .lock()
+            .unwrap();
+        state.reset();
+        state.current_slot = slot;
+        state.current_slot_start = Instant::now();
+        for (peer, &stake) in staked_nodes.iter() {
+            let Some(ag_addr) = self
+                .cluster_info
+                .lookup_contact_info(peer, |ci| ci.alpenglow())
+                .flatten()
+            else {
+                continue;
+            };
+            state.peers.insert(
+                *peer,
+                PeerData {
+                    stake: stake,
+                    address: ag_addr,
+                    relative_toa: [None; 4],
+                },
+            );
+            state.total_staked += stake;
+        }
+        trace!(
+            "Prepared for slot {slot}, total stake is {}",
+            state.total_staked
+        );
+    }
+
+    /// Collects votes and changes states
+    fn listener_thread(
+        self_state: Arc<[Mutex<SharedState>; 3]>,
+        should_exit: Arc<AtomicBool>,
+        cluster_info: Arc<ClusterInfo>,
+        mut epoch_specs: EpochSpecs,
+        socket: Arc<UdpSocket>,
+        command_sender: Sender<Command>,
+    ) {
+        socket
+            .set_read_timeout(Some(Duration::from_secs(1)))
+            .unwrap();
+        let id = cluster_info.id();
+        // Set aside enough space to fetch multiple packets from the kernel per syscall
+        let mut packets: Vec<Packet> = vec![Packet::default(); 1024];
+        loop {
+            // must wipe all Meta records to reuse the buffer
+            for p in packets.iter_mut() {
+                *p.meta_mut() = Meta::default();
+            }
+            // recv_mmsg should timeout in 1 second
+            let n = match recv_mmsg(&socket, &mut packets) {
+                Ok(n) => n,
+                Err(e) => {
+                    match e.kind() {
+                        std::io::ErrorKind::TimedOut | std::io::ErrorKind::WouldBlock => {
+                            0 // no packets received
+                        }
+                        _ => {
+                            error!("Got error {:?} in mock alpenglow RX socket operation, exiting thread", e.raw_os_error());
+                            return;
+                        }
+                    }
+                }
+            };
+            if should_exit.load(Ordering::Relaxed) {
+                return;
+            }
+            // we may have received no packets, in this case we can safely skip the rest
+            if n == 0 {
+                continue;
+            }
+
+            for pkt in packets.iter().take(n) {
+                if pkt.meta().size < MOCK_VOTE_HEADER_SIZE {
```

Comment:
should we check on a specific size? All of these packets should be exactly the same size, right?